File: csharp/1838-Frequency-Of-The-Most-Frequent-Element.cs
public class Solution {
    public int MaxFrequency(int[] nums, int k) {
        Array.Sort(nums);
        
        int left = 0, right = 0;
        var max = 0;
        long totalSum = 0;
        
        for(right = 0; right < nums.Length; right++) {
            totalSum += nums[right];
            
            while(nums[right] * (right - left + 1) > totalSum + k) {
                totalSum -= nums[left];
                left++;
            }
            
            max = Math.Max((right - left + 1), max);
        }
        
        return max;
    }
}

File: java/1838-frequency-of-the-most-frequent-element.java
class Solution {
    public int maxFrequency(int[] nums, int k) {
        Arrays.sort(nums);
        int left = 0, currSum = 0, res = 0;

        for(int right = 0; right < nums.length; right++){
            int target = nums[right];
            currSum += nums[right];

            while((right - left + 1)*target - currSum > k){
                currSum -= nums[left];
                left++;
            }
            res = Math.max(res, right - left + 1);
        }
        return res;
    }
}
